include/seqan3/alignment/matrix/detail/coordinate_matrix.hpp

```diff
@@ -12,43 +12,172 @@
+#include <seqan3/range/container/aligned_allocator.hpp>
 
 namespace seqan3::detail
 {
 
-/*!\brief A function object that converts a column index and a row index to a seqan3::detail::matrix_coordinate.
+//------------------------------------------------------------------------------
+// coordinate_matrix
+//------------------------------------------------------------------------------
+
+/*!\brief A matrix over coordinates.
  * \ingroup alignment_matrix
+ * \implements std::ranges::forward_range
+ *
+ * \tparam index_t The underlying matrix index type; must mode std::integral or seqan3::simd::simd_concept with the
+ *                 packed scalar type modeling std::integral.
  *
  * \details
  *
- * This helper function object is used for the seqan3::detail::coordinate_matrix_iterator to convert a column index
- * and row index into a seqan3::detail::matrix_coordinate. The seqan3::detail::coordinate_matrix_iterator iterates
- * over the columns of the underlying coordinate matrix and returns a transformed iota range over the row indices of
- * this column. The column index is stored as member inside of this function object since every column index is
- * associated with many row indices.
+ * This matrix emulates a two-dimensional index, such that each cell inside of the alignment matrix can be located
+ * through a unique coordinate (=index). In the alignment algorithm this matrix is paired with the alignment matrix
+ * to form an indexed alignment matrix.
+ *
+ * This matrix is cheap as it stores only the dimensions of the matrix and does not allocate any memory for the
+ * coordinates. It uses the seqan3::detail::matrix_coordinate_iterator to iterate
+ * over the columns of this virtual matrix. When the seqan3::detail::coordinate_matrix::matrix_coordinate_iterator
+ * is dereferenced it returns an on-the-fly constructed range
+ * ([prvalue](https://en.cppreference.com/w/cpp/language/value_category))
+ * representing the current indexed column. The reference type of this range is seqan3::detail::matrix_coordinate.
+ *
+ * ### Simd mode
+ *
+ * If the `index_t` is a simd vector the matrix implements a vectorised coordinate matrix instead. In this case the
+ * matrix uses a seqan3::detail::coordinate_matrix::matrix_coordinate_simd_iterator to iterate over the vectorised
+ * matrix. When the seqan3::detail::coordinate_matrix::matrix_coordinate_iterator is dereferenced it returns an
+ * on-the-fly constructed range ([prvalue](https://en.cppreference.com/w/cpp/language/value_category)) representing the
+ * current simd coordinate column. The reference type of this range is seqan3::detail::simd_matrix_coordinate.
  */
-struct convert_to_matrix_coordinate
+template <typename index_t>
+//!\cond
+    requires std::integral<index_t> ||
+             (simd::simd_concept<index_t> && requires (simd::simd_traits<index_t> traits)
+             {
+                 requires std::integral<typename decltype(traits)::scalar_type>;
+             })
```

Comment:
```suggestion
    requires (std::integral<index_t> ||
              (simd::simd_concept<index_t> && requires ()
              {
                  requires std::integral<typename simd::simd_traits<index_t>::scalar_type>;
              }))
```

Reply:
This is already use in two-places, maybe we should make thus a named concept. How about `scalar_or_simd_index`.